File: articles/population-dynamics-ecological-modeling/python/stochastic_population_viability.py
"""
Stochastic population viability simulation.

This script simulates logistic growth with variable growth rate, variable
carrying capacity, harvest, catastrophes, and quasi-extinction screening.

Run:
    python python/stochastic_population_viability.py
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]
SCENARIO_PATH = ARTICLE_DIR / "data" / "population_scenarios.csv"


def simulate_population(
    years: int,
    n_sims: int,
    initial_population: float,
    growth_rate_mean: float,
    growth_rate_sd: float,
    carrying_capacity_mean: float,
    carrying_capacity_sd: float,
    harvest: float,
    catastrophe_probability: float,
    catastrophe_multiplier: float,
    quasi_extinction_threshold: float,
    random_seed: int = 123,
) -> dict[str, float]:
    """Run stochastic population viability simulation."""

    rng = np.random.default_rng(random_seed)
    trajectories = np.full((years + 1, n_sims), np.nan)
    trajectories[0, :] = initial_population

    for sim in range(n_sims):
        population_size = initial_population

        for year in range(1, years + 1):
            growth_rate_t = rng.normal(growth_rate_mean, growth_rate_sd)
            carrying_capacity_t = max(
                quasi_extinction_threshold,
                rng.normal(carrying_capacity_mean, carrying_capacity_sd),
            )

            population_size = (
                population_size
                + growth_rate_t
                * population_size
                * (1.0 - population_size / carrying_capacity_t)
                - harvest
            )

            if rng.random() < catastrophe_probability:
                population_size *= catastrophe_multiplier

            population_size = max(0.0, population_size)
            trajectories[year, sim] = population_size

            if population_size == 0.0:
                trajectories[year:, sim] = 0.0
                break

    final_sizes = trajectories[years, :]
    minimum_sizes = np.nanmin(trajectories, axis=0)

    return {
        "extinction_risk": float(np.mean(final_sizes == 0.0)),
        "quasi_extinction_risk": float(
            np.mean(minimum_sizes <= quasi_extinction_threshold)
        ),
        "mean_final": float(np.mean(final_sizes)),
        "median_final": float(np.median(final_sizes)),
    }


def main() -> None:
    """Compare population viability across scenarios."""

    scenarios = pd.read_csv(SCENARIO_PATH)

    rows = []
    for _, row in scenarios.iterrows():
        result = simulate_population(
            years=50,
            n_sims=1000,
            initial_population=row["initial_population"],
            growth_rate_mean=row["growth_rate_mean"],
            growth_rate_sd=row["growth_rate_sd"],
            carrying_capacity_mean=row["carrying_capacity_mean"],
            carrying_capacity_sd=row["carrying_capacity_sd"],
            harvest=row["harvest"],
            catastrophe_probability=row["catastrophe_probability"],
            catastrophe_multiplier=row["catastrophe_multiplier"],
            quasi_extinction_threshold=row["quasi_extinction_threshold"],
        )
        result["scenario"] = row["scenario"]
        rows.append(result)

    output = pd.DataFrame(rows)[
        [
            "scenario",
            "extinction_risk",
            "quasi_extinction_risk",
            "mean_final",
            "median_final",
        ]
    ]

    print(output.round(3).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/population-dynamics-ecological-modeling/cpp/stochastic_population_viability.cpp
/*
 * Stochastic population viability simulation in C++.
 *
 * This compact example simulates logistic growth with harvest, environmental
 * variability, and catastrophe risk.
 */

#include <algorithm>
#include <iostream>
#include <random>
#include <vector>

int main() {
    const int years = 50;
    const int n_sims = 1000;

    const double initial_population = 80.0;
    const double growth_rate_mean = 0.18;
    const double growth_rate_sd = 0.08;
    const double carrying_capacity_mean = 500.0;
    const double carrying_capacity_sd = 40.0;
    const double harvest = 5.0;
    const double catastrophe_probability = 0.05;
    const double catastrophe_multiplier = 0.60;
    const double quasi_extinction_threshold = 20.0;

    std::mt19937 rng(123);
    std::normal_distribution<double> growth_dist(growth_rate_mean, growth_rate_sd);
    std::normal_distribution<double> capacity_dist(carrying_capacity_mean, carrying_capacity_sd);
    std::uniform_real_distribution<double> uniform(0.0, 1.0);

    int extinct_count = 0;
    int quasi_extinct_count = 0;
    double final_sum = 0.0;

    for (int sim = 0; sim < n_sims; ++sim) {
        double population_size = initial_population;
        double minimum_size = population_size;

        for (int year = 0; year < years; ++year) {
            double growth_rate_t = growth_dist(rng);
            double carrying_capacity_t = std::max(
                quasi_extinction_threshold,
                capacity_dist(rng)
            );

            population_size =
                population_size +
                growth_rate_t * population_size *
                (1.0 - population_size / carrying_capacity_t) -
                harvest;

            if (uniform(rng) < catastrophe_probability) {
                population_size *= catastrophe_multiplier;
            }

            population_size = std::max(0.0, population_size);
            minimum_size = std::min(minimum_size, population_size);

            if (population_size == 0.0) {
                break;
            }
        }

        if (population_size == 0.0) {
            extinct_count += 1;
        }

        if (minimum_size <= quasi_extinction_threshold) {
            quasi_extinct_count += 1;
        }

        final_sum += population_size;
    }

    std::cout << "extinction_risk=" << static_cast<double>(extinct_count) / n_sims << std::endl;
    std::cout << "quasi_extinction_risk=" << static_cast<double>(quasi_extinct_count) / n_sims << std::endl;
    std::cout << "mean_final_population=" << final_sum / n_sims << std::endl;

    return 0;
}
